File: backend/src/routes/prices.ts
import { Router } from 'express';
import { supabaseAdmin } from '../lib/supabase';
import type { Prices } from '../types';

export const pricesRouter = Router();

const CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes

interface PriceCacheRow {
  coin_id: string;
  price_brl: number;
  image_url: string | null;
  updated_at: string;
}

async function fetchFromCoinGecko(ids: string[]): Promise<{ id: string; price: number; image?: string }[]> {
  const apiKey = process.env.COINGECKO_API_KEY;
  const keyParam = apiKey ? `&x_cg_demo_api_key=${apiKey}` : '';
  const url = `https://api.coingecko.com/api/v3/coins/markets?vs_currency=brl&ids=${ids.join(',')}${keyParam}`;

  const r = await fetch(url);
  if (!r.ok) throw Object.assign(new Error('coingecko_http_error'), { status: r.status });
  const d = await r.json();
  if (!Array.isArray(d)) throw new Error('coingecko_invalid_response');

  return d
    .filter((coin) => coin.current_price != null)
    .map((coin) => ({ id: coin.id as string, price: coin.current_price as number, image: coin.image as string | undefined }));
}

// GET /api/prices?ids=bitcoin,ethereum — 5-minute cache in the price_cache table
pricesRouter.get('/', async (req, res) => {
  const idsParam = (req.query.ids as string | undefined)?.trim();
  if (!idsParam) {
    res.status(400).json({ error: 'Query param "ids" is required (comma-separated).' });
    return;
  }
  const ids = [...new Set(idsParam.split(',').map((s) => s.trim()).filter(Boolean))];

  const { data: cachedRows, error: cacheError } = await supabaseAdmin
    .from('price_cache')
    .select('*')
    .in('coin_id', ids);

  if (cacheError) {
    res.status(500).json({ error: cacheError.message });
    return;
  }

  const now = Date.now();
  const fresh = new Map<string, PriceCacheRow>();
  for (const row of (cachedRows ?? []) as PriceCacheRow[]) {
    if (now - new Date(row.updated_at).getTime() < CACHE_TTL_MS) {
      fresh.set(row.coin_id, row);
    }
  }

  const staleIds = ids.filter((id) => !fresh.has(id));
  const result: Prices = {};
  for (const [id, row] of fresh) result[id] = Number(row.price_brl);

  if (staleIds.length > 0) {
    try {
      const fetched = await fetchFromCoinGecko(staleIds);
      if (fetched.length > 0) {
        await supabaseAdmin.from('price_cache').upsert(
          fetched.map((c) => ({ coin_id: c.id, price_brl: c.price, image_url: c.image ?? null, updated_at: new Date().toISOString() }))
        );
      }
      for (const c of fetched) result[c.id] = c.price;
    } catch (e) {
      // If CoinGecko fails, still return whatever we had cached (even if
      // stale) instead of failing the whole response.
      for (const id of staleIds) {
        const stale = (cachedRows as PriceCacheRow[] | null)?.find((r) => r.coin_id === id);
        if (stale) result[id] = Number(stale.price_brl);
      }
      const status = (e as { status?: number }).status === 429 ? 429 : undefined;
      if (Object.keys(result).length === 0) {
        res.status(status ?? 502).json({ error: 'Failed to fetch prices from CoinGecko.' });
        return;
      }
    }
  }

  res.json(result);
});

File: backend/app/routes/prices.py
import datetime
import re
import time
from fastapi import APIRouter, Depends, HTTPException, Query, status
from app.dependencies import require_auth
from app.db.postgres_client import get_conn
from app.models import PriceInfo
from app.price_provider import get_provider, PricedAsset

router = APIRouter()

_CACHE_TTL_S = 5 * 60
_COIN_ID_RE = re.compile(r'^[a-z0-9-]{1,120}$')


@router.get("", response_model=dict[str, PriceInfo])
def get_prices(
    ids: str = Query(..., description="Comma-separated CoinGecko coin IDs"),
    _auth=Depends(require_auth),
):
    coin_ids = list({s.strip() for s in ids.split(",") if s.strip()})
    if not coin_ids:
        raise HTTPException(status_code=400, detail='Query param "ids" is required.')

    invalid = [cid for cid in coin_ids if not _COIN_ID_RE.fullmatch(cid)]
    if invalid:
        raise HTTPException(status_code=400, detail=f"Invalid coin_id(s): {', '.join(invalid)}")

    conn = get_conn()

    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT coin_id, price_usd, image_url, updated_at FROM price_cache"
                " WHERE coin_id = ANY(%s)",
                (coin_ids,),
            )
            cached = cur.fetchall()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

    now = time.time()
    fresh: dict[str, dict] = {}
    for row in cached:
        updated = row["updated_at"]
        ts = updated.timestamp() if hasattr(updated, "timestamp") else datetime.datetime.fromisoformat(str(updated).replace("Z", "+00:00")).timestamp()
        if now - ts < _CACHE_TTL_S:
            fresh[row["coin_id"]] = row

    stale_ids = [cid for cid in coin_ids if cid not in fresh]
    result: dict[str, PriceInfo] = {
        cid: PriceInfo(price=float(row["price_usd"]), image=row.get("image_url"))
        for cid, row in fresh.items()
    }

    if stale_ids:
        try:
            fetched = get_provider().get_prices([PricedAsset(coin_id=cid) for cid in stale_ids])
            if fetched:
                now_iso = datetime.datetime.now(datetime.timezone.utc).isoformat()
                with conn.cursor() as cur:
                    cur.executemany(
                        "INSERT INTO price_cache (coin_id, price_usd, image_url, updated_at)"
                        " VALUES (%s, %s, %s, %s)"
                        " ON CONFLICT (coin_id) DO UPDATE SET price_usd = EXCLUDED.price_usd,"
                        " image_url = EXCLUDED.image_url, updated_at = EXCLUDED.updated_at",
                        [(c["id"], c["price"], c.get("image"), now_iso) for c in fetched],
                    )
                conn.commit()
            for c in fetched:
                result[c["id"]] = PriceInfo(price=c["price"], image=c.get("image"))
        except NotImplementedError as e:
            # Must precede `except Exception` below: a provider that doesn't implement
            # this capability is a permanent condition, not a transient upstream hiccup,
            # so it must never be treated as stale-cache-fallback material.
            raise HTTPException(status_code=501, detail=str(e))
        except HTTPException:
            # On CoinGecko failure, fall back to stale cache rather than erroring
            for row in cached:
                if row["coin_id"] in stale_ids and row["coin_id"] not in result:
                    result[row["coin_id"]] = PriceInfo(price=float(row["price_usd"]), image=row.get("image_url"))
            if not result:
                raise
        except Exception as e:
            conn.rollback()
            for row in cached:
                if row["coin_id"] in stale_ids and row["coin_id"] not in result:
                    result[row["coin_id"]] = PriceInfo(price=float(row["price_usd"]), image=row.get("image_url"))
            if not result:
                raise HTTPException(status_code=502, detail=str(e))

    return result
